includes/discount-functions.php

```diff
@@ -373,31 +373,52 @@ function edd_get_discount_notes( $discount_id = 0 ) {
  *
  * @since 1.0
  * @since 3.0 Updated to be more efficient and make direct calls to the EDD_Discount object.
+ * @since 3.1.0.3 Added Sample size and Refresh parameters to be able to cache the result as a transient.
+ *
+ * @param int  $sample_size The number of discount codes to query. Defaults to 10 for performance.
+ * @param bool $refresh     If we should force a refresh of the transient. Defaults to false.
  *
  * @return bool
  */
-function edd_has_active_discounts() {
+function edd_has_active_discounts( $sample_size = 10, $refresh = false ) {
 
-	// Query for active discounts.
-	$discounts = edd_get_discounts( array(
-		'number' => 10,
-		'status' => 'active'
-	) );
+	// If we are not forcing a refresh of the value, check the transient.
+	if ( false === $refresh ) {
+		$cached_result = get_transient( 'edd_has_active_discounts' );
 
-	// Bail if none.
-	if ( empty( $discounts ) ) {
-		return false;
+		if ( false !== $cached_result ) {
+			return boolval( $cached_result );
+		}
 	}
 
-	// Check each discount for active status, applying filters, etc...
-	foreach ( $discounts as $discount ) {
-		/** @var $discount EDD_Discount */
-		if ( $discount->is_active( false, true ) ) {
-			return true;
+	$has_active_discounts = false;
+
+	// Query for active discounts.
+	$discounts = edd_get_discounts(
```

Comment:
This is better, but I'm not sure it's fixing the issue as defined in the replication steps.

This is querying the ten most recent `active` discounts, in this case, and if they're all expired, then the function will return false. I experimented with updating the discounts query to include an end date query, something like this (I know this is hard coded so would obviously have to change):
```
	// Query for active discounts.
	$discounts = edd_get_discounts(
		array(
			'number'         => $sample_size,
			'status'         => 'active',
			'end_date_query' => array(
				'relation' => 'OR',
				array(
					'after'     => array(
						'year'  => 2022,
						'month' => 11,
						'day'   => 16,
					),
					'inclusive' => false,
				),
			),
		)
	);
```
That worked to only get active discounts with future expiration dates, but it fails to get active discounts with no expiration dates, and looking at the `WP_Date_Query` class, I am not sure how to do a combination query for a `null` value in the expiration column. But it seems like if we could do this query, it would solve the reported issue.